File: Data Structures & Algorithms/products-of-array-discluding-self/submission-1.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int cur = 1;
        int[] res = new int[nums.length];
        int[] pre = new int[nums.length];
        int[] suf = new int[nums.length];

        for (int i = 0 ; i < nums.length ; i++){
            pre[i] = cur;
            cur *= nums[i];
        }

        cur = 1;
        for (int i = nums.length - 1 ; i >= 0 ; i--) {
            suf[i] = cur;
            cur *= nums[i];
        }

        for (int i = 0 ; i < nums.length ; i++) res[i] = suf[i] * pre[i];
        
        return res;
    }
}  

File: Data Structures & Algorithms/generate-parentheses/submission-1.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        # My thinking
        # Lets say n = 3
        # 1: (()())
        stack = []
        res = []

        def backtrack(cur: str, l: int, r: int):
            if l == r == n:
                res.append(cur)
                return

            if l < n:
                backtrack(cur + "(", l + 1, r)
            if r < l:
                backtrack(cur + ")", l, r + 1)     

        backtrack("", 0,0)    

        return res